### DISCUSSION TITLE
My 3 home-made Programmer

### DISCUSSION THREAD

**Discussion by flyandancexo** [avrdudes/avrdude#1693] [Show and tell]
A photo worth thousands words, guess I share few photos here:
![FDxICSP1_M](https://github.com/avrdudes/avrdude/assets/66555404/b257bb19-b7e1-4b71-9694-85a117e71609)

![FDxICSP1_B](https://github.com/avrdudes/avrdude/assets/66555404/66d8af23-1187-4b50-983b-431378abc235)
![FDxICSP1_C](https://github.com/avrdudes/avrdude/assets/66555404/966b2714-5c98-4513-93be-d04d960afbb5)
![FDxICSP1_D](https://github.com/avrdudes/avrdude/assets/66555404/baa2276a-90c9-46a6-b830-482f804850a5)
![FDxICSP1_E](https://github.com/avrdudes/avrdude/assets/66555404/9fca09bc-edb2-4353-9b40-d174ea1843c7)
![FDxICSP1_F](https://github.com/avrdudes/avrdude/assets/66555404/650aa3b2-ae5c-4f24-9a84-3cb1371c977c)
![FDxICSP1_H](https://github.com/avrdudes/avrdude/assets/66555404/12ed55a5-3a4a-467b-8665-f91d68c59d2e)
![FDxICSP1_I](https://github.com/avrdudes/avrdude/assets/66555404/36aca21b-0e02-432e-8ba8-34bd1a195360)
![FDxICSP1_J](https://github.com/avrdudes/avrdude/assets/66555404/d055c301-04e4-4c52-849e-4cadf5965d5b)
![FDxICSP1_K](https://github.com/avrdudes/avrdude/assets/66555404/9b805b38-d5f0-4b55-aaf8-81c1664b0645)

FDxICSP v1.81 Upload Screenshot: (Avrdude v7.2)
FDxICSP_V181_SCK_2MHz_m8:
![FDxICSP_V181_SCK_2MHz_m8](https://github.com/avrdudes/avrdude/assets/66555404/0524280d-d932-4131-94cf-29b28e9c0d8f)

FDxICSP_V181_SCK_2MHz_m88p:
![FDxICSP_V181_SCK_2MHz_m88p](https://github.com/avrdudes/avrdude/assets/66555404/f75c3c1f-bfac-4181-975e-ecefc80a2b47)

FDxICSP_V181_SCK_2MHz_m128:
![FDxICSP_V181_SCK_2MHz_m128](https://github.com/avrdudes/avrdude/assets/66555404/6e2765b4-5316-4777-8938-9579e8a89ccd)

FDxICSP_V181_SCK_2MHz_m328p:
![FDxICSP_V181_SCK_2MHz_m328p](https://github.com/avrdudes/avrdude/assets/66555404/20eec96f-16e1-421a-9916-305bf60b8533)

FDxICSP_V181_SCK_4MHz_m128:
![FDxICSP_V181_SCK_4MHz_m128](https://github.com/avrdudes/avrdude/assets/66555404/cacbe69a-4ada-4e21-8d73-c93852be3243)

FDxICSP_V181_SCK_4MHz_m328p:
![FDxICSP_V181_SCK_4MHz_m328p](https://github.com/avrdudes/avrdude/assets/66555404/a7a4d459-c9fa-450e-a235-3e115e223863)

**Comment by mcuee** [comment]
Nice PCBA build.

For the Flash read/write speed benchmark, you may want to post in the other discussion as well.
* https://github.com/avrdudes/avrdude/discussions/1022

**Reply by flyandancexo** [reply]
I read that already. I don't want to brag about how fast my programmer is. You should also start a post for benchmark bootloader too. It could be very interested to know the upload speed of current bootloaders.

**Reply by mcuee** [reply]
You can post results for bootloaders in the same discussion. It is not limited to programmers.
* https://github.com/avrdudes/avrdude/discussions/1022

**Reply by mcuee** [reply]
@flyandancexo 

Bootloader test results -- urboot and ATmega328P
* https://github.com/avrdudes/avrdude/discussions/1022#discussioncomment-8537894

**Comment by stefanrueger** [comment]
> You can post results for bootloaders in the same discussion. It is not limited to programmers.

Correct, but note that we prefer info on *published* bootloaders, ie, ones that peeps can actually use.

**Reply by flyandancexo** [reply]
I am sorry my friend. I am struggling on everything including getting my bootloader done, but I will publish it when it's done. Here is my new board that I will use to develop the bootloader and other stuff. I am really trying very hard and my best while being in a state of struggling. 2Mbps is still not reading back correctly, probably CH340g is not stable enough at 2Mbps. Meanwhile, can you upload the test results of other published bootloader, so I can have an idea what other bootloader is doing and try to do better. Thank you.
![A](https://github.com/avrdudes/avrdude/assets/66555404/81e037b2-ac4b-4f70-b488-3065d5c6c494)

**Reply by mcuee** [reply]
CH340 may not be the best adapter to go if you are chasing for speed. That being said, 2Mbps may not be much faster than to 500Kbps for AVR chip Flash read/write.

Reference: the following is talking about serialupdi using different USB to TTL adapter, just for your reference. CH340x is cheap but may not have the best performance.
https://github.com/SpenceKonde/AVR-Guidance/blob/master/UPDI/jtag2updi.md

**Reply by flyandancexo** [reply]
CH340g is great. I have been using it for decade. It's not the issue, and here is my calculation and test. 

```
UART : At 1Mbps it takes 1/(1000000/((2+8)*(256+3+4)))=0.00263s to transfer 256 bytes
SPI  : At 4Mhz, it takes 1/(4000000/(256*4*8))=0.002048s to transfer that same 256
SPI  : At 62khz, it takes 1/(62000/(256*4*8))=0.132s to transfer that same 256
Flash: It takes 3.7-4.5ms to write to 1 page or 256 bytes
```

Total time for SPI to write one page = 0.002048+0.0045 = 0.006548
Absolute maximum write speed       = 256/0.006548=39095


Here are my record of BAUD VS upload speed for older version:
```
1Mbps:
4M - 7162/0.43=16655
2M - 7162/0.49=14616
1M - 7162/0.62=11551
500K 7162/0.86=8327
250K 7162/1.34=5344
125K 7162/2.32=3087
62K  7162/4.19=1709

500Kbps:
4M - 7162/0.51=14043
2M - 7162/0.56=12789
1M - 7162/0.68=10532
500K 7162/0.92=7784
250K 7162/1.42=5043
125K 7162/2.39=2996
62K  7162/4.26=1681

250Kbps:
4M - 7162/0.66=10851
2M - 7162/0.71=10087
1M - 7162/0.84=8526
500K 7162/1.08=6631
250K 7162/1.58=4532
125K 7162/2.54=2819
62K  7162/4.41=1624
```

**Comment by mcuee** [comment]
@flyandancexo 

I think your AVR910 is quite optimized and indeed the speed is quite fast.

Just FYI the fastest I see so far is from FTDI FT232H/FTDI2232H JTAG (8MHz): 128KB Flash reading at 0.40 seconds.
* https://github.com/avrdudes/avrdude/pull/1324

```
$ avrdude -p atmega128rfa1 -c tumpa-jtag -B 8MHz -U flash:r:dump

avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1ea701 (probably m128rfa1)
avrdude: reading flash memory ...

Reading | ################################################## | 100% 0.40 s 

avrdude: writing output file dump

avrdude done.  Thank you.
```

**Reply by flyandancexo** [reply]
This is fast, but my test is done using 4MHz SCK, so it's not really a fair comparison.

**Comment by mcuee** [comment]
More info here: no firmware source codes available.
https://github.com/flyandancexo/FDxICSP
